olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -125,6 +143,57 @@ http::NetworkRequest::HttpVerb GetHttpVerb(const std::string& verb) {
   return http_verb;
 }
 
+HttpResponse SendRequest(const http::NetworkRequest& request,
+                         const olp::client::OlpClientSettings& settings,
+                         const olp::client::RetrySettings& retry_settings,
+                         client::CancellationContext context) {
+  http::NetworkResponse network_response = kCancelledErrorResponse;
+  auto interest_flag = std::make_shared<std::atomic_bool>(true);
+  Condition condition{};
+  auto response_body = std::make_shared<std::stringstream>();
+  http::SendOutcome outcome{http::ErrorCode::CANCELLED_ERROR};
+
+  context.ExecuteOrCancelled(
+      [&]() {
+        outcome = settings.network_request_handler->Send(
+            request, response_body,
+            [&, interest_flag](http::NetworkResponse response) {
+              if (interest_flag->exchange(false)) {
+                network_response = std::move(response);
+                condition.Notify();
+              }
+            });
+
+        return CancellationToken([&, interest_flag]() {
+          if (interest_flag->exchange(false)) {
+            settings.network_request_handler->Cancel(outcome.GetRequestId());
+            network_response = kCancelledErrorResponse;
+            network_response.WithRequestId(outcome.GetRequestId());
+            condition.Notify();
+          }
+        });
+      },
+      [&condition]() { condition.Notify(); });
+
+  if (!outcome.IsSuccessful()) {
+    return {static_cast<int>(outcome.GetErrorCode()),
+            ErrorCodeToString(outcome.GetErrorCode())};
+  }
+
+  if (!condition.Wait(std::chrono::seconds(retry_settings.timeout))) {
+    OLP_SDK_LOG_INFO_F(kLogTag, "Timeout");
+    context.CancelOperation();
+    return ToHttpResponse(kTimeoutErrorResponse);
+  }
+
+  if (context.IsCancelled()) {
```

Comment:
We need to exchange  interest_flag to false here.

Reply:
If request was cancelled before execution, we reach here without calling callback or cancellation token, meaning interest_flag is true, but no `Send` called and no `CancellationToken` with local variables created, so we can safely escape. Add test `OlpClientTest.CancelBeforeExecution` to cover such case.